File: backend/src/main/java/com/ll/hotel/domain/hotel/hotel/dto/HotelDetailDto.java
package com.ll.hotel.domain.hotel.hotel.dto;

import com.ll.hotel.domain.hotel.hotel.entity.Hotel;
import com.ll.hotel.domain.hotel.option.entity.HotelOption;
import com.ll.hotel.domain.hotel.room.dto.GetRoomResponse;
import com.ll.hotel.domain.hotel.room.dto.RoomWithImageDto;
import jakarta.validation.constraints.NotBlank;
import java.time.LocalTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.NonNull;

public record HotelDetailDto(
        long hotelId,

        @NotBlank
        String hotelName,

        @NotBlank
        String hotelEmail,

        @NotBlank
        String hotelPhoneNumber,

        @NotBlank
        String streetAddress,

        @NonNull
        Integer zipCode,

        @NonNull
        Integer hotelGrade,

        @NonNull
        LocalTime checkInTime,

        @NonNull
        LocalTime checkOutTime,

        @NotBlank
        String hotelExplainContent,

        @NotBlank
        String hotelStatus,

        List<GetRoomResponse> rooms,

        Set<String> hotelOptions
) {
    public HotelDetailDto(Hotel hotel, List<RoomWithImageDto> dtos) {
        this(
                hotel.getId(),
                hotel.getHotelName(),
                hotel.getHotelEmail(),
                hotel.getHotelPhoneNumber(),
                hotel.getStreetAddress(),
                hotel.getZipCode(),
                hotel.getHotelGrade(),
                hotel.getCheckInTime(),
                hotel.getCheckOutTime(),
                hotel.getHotelExplainContent(),
                hotel.getHotelStatus().name(),
                dtos.stream()
                        .map(GetRoomResponse::new)
                        .toList(),
                hotel.getHotelOptions() != null
                ? hotel.getHotelOptions().stream()
                        .map(HotelOption::getName)
                        .collect(Collectors.toSet())
                        : new HashSet<>()
        );
    }
}

File: backend/src/main/java/com/ll/hotel/domain/hotel/hotel/dto/HotelDetailDto.kt
package com.ll.hotel.domain.hotel.hotel.dto

import com.ll.hotel.domain.hotel.hotel.entity.Hotel
import com.ll.hotel.domain.hotel.room.dto.GetRoomResponse
import com.ll.hotel.domain.hotel.room.dto.RoomWithImageDto
import jakarta.validation.constraints.NotBlank
import java.time.LocalTime

data class HotelDetailDto(
    val hotelId: Long,

    @field:NotBlank
    val hotelName: String,

    @field:NotBlank
    val hotelEmail: String,

    @field:NotBlank
    val hotelPhoneNumber: String,

    @field:NotBlank
    val streetAddress: String,

    val zipCode: Int,

    val hotelGrade: Int,

    val checkInTime: LocalTime,

    val checkOutTime: LocalTime,

    @field:NotBlank
    val hotelExplainContent: String,

    @field:NotBlank
    val hotelStatus: String,

    val rooms: List<GetRoomResponse>,

    val hotelOptions: Set<String>
) {
    constructor(hotel: Hotel, dtos: List<RoomWithImageDto>) : this(
        hotel.id,
        hotel.hotelName,
        hotel.hotelEmail,
        hotel.hotelPhoneNumber,
        hotel.streetAddress,
        hotel.zipCode,
        hotel.hotelGrade,
        hotel.checkInTime,
        hotel.checkOutTime,
        hotel.hotelExplainContent,
        hotel.hotelStatus.name,
        dtos.map { GetRoomResponse(it) },
        hotel.hotelOptions?.takeIf { it.isNotEmpty() }?.map { it.name }?.toSet() ?: emptySet()
    )
}